src/main/kotlin/org/rust/cargo/project/settings/RustProjectSettingsService.kt

```diff
@@ -61,104 +88,55 @@ interface RustProjectSettingsService {
         fun setToolchain(toolchain: RsToolchain?) {
             toolchainHomeDirectory = toolchain?.location?.systemIndependentPath
         }
-    }
 
-    enum class MacroExpansionEngine {
-        DISABLED,
-        OLD, // `OLD` can't be selected by a user anymore, it exists for backcompat with saved user settings
-        NEW
+        override fun copy(): State {
+            val state = State()
+            state.copyFrom(this)
+            return state
+        }
     }
 
-    @Retention(AnnotationRetention.RUNTIME)
-    @Target(AnnotationTarget.PROPERTY)
-    private annotation class AffectsCargoMetadata
+    override fun loadState(state: State) {
+        if (state.macroExpansionEngine == MacroExpansionEngine.OLD) {
+            state.macroExpansionEngine = MacroExpansionEngine.NEW
+        }
+        super.loadState(state)
+    }
 
-    @Retention(AnnotationRetention.RUNTIME)
-    @Target(AnnotationTarget.PROPERTY)
-    private annotation class AffectsHighlighting
+    override fun notifySettingsChanged(event: SettingsChangedEventBase<State>) {
+        super.notifySettingsChanged(event)
 
-    /**
-     * Allows to modify settings.
-     * After setting change,
-     */
-    fun modify(action: (State) -> Unit)
+        if (event.isChanged(State::doctestInjectionEnabled)) {
+            // flush injection cache
+            PsiManager.getInstance(project).dropPsiCaches()
+        }
```

Comment:
Why did you extract this check into `RustProjectSettingsService` instead of keeping it in base class?